### PR TITLE
Add /preimage_received endPoint and update preimage

### LINKED ISSUE #52: Implement /preimage_received endpoint
API to implement:

```
POST /preimage_received
```

The node periodically pushes the `preimage`.
`Stoa` does receive the preimage and store it in the `Stoa database`.
`The preimage` saves only the last `preimage` close to `cycle_length` for each `PreImageCycle`.
This can enhancement `the preimage` Infomation of `/getValidator` and `/getValidators`.

JSON Type of `preimageInfo`

```json
{
    "enroll_key":"0x000000000019d6689c085ae",
    "hash":"0x4a5e1e4baab89f3a32518a88",
    "distance":6
}
```

### PR DESCRIPTION
**Pull request by TrustHenry**
Receive the preimageInfo and store the preimage information in the validators table.
This saves only new preimage information and the distance must be greater than the previously saved one.
And the distance cannot be greater than the cycle of the enrollment.
The preimage information is provided in /getvalidator /getvalidators .
Returns the preimage of the enrolled Enrollment by the Height value.
At this point, the preimage can be found by hashing the preimage to the distance.
The Validators table has the last released preimage.
And it's each stored on a continuous Enrollment cycle.

Related to #52

### CODE REVIEW

File: tests/Storage.test.ts [hunk 1]
```diff
@@ -251,3 +256,86 @@ describe('LedgerStorage', () =>
...
 });
+
+/**
+ *  Run the test of the update preImage
+ */
+function updatePreImageTest (ledger_storage: LedgerStorage, onDone: () => void)
+{
+    let height: number = 0;
+    ledger_storage.getEnrollments(height,
+        (rows: any[]) =>
```
Comment (MichaelKim20) [inline tests/Storage.test.ts]: Why do you need this? And overlaps with the 'rows' internal declaration.

File: tests/Storage.test.ts [hunk 2]
```diff
@@ -251,3 +256,86 @@ describe('LedgerStorage', () =>
...
+        {
```
Comment (MichaelKim20) [inline tests/Storage.test.ts:366]: Since the three internal tests are run simultaneously, the remaining tests are not run normally by the tests that end first.

File: src/Stoa.ts [hunk 3]
```diff
@@ -45,15 +45,30 @@ class Stoa {
...
                         for (const row of rows)
                         {
-                            let preimage: IPreimage = {distance: row.distance,
-                                hash: (row.distance == 0 ? row.random_seed : '')} as IPreimage;
+                            let preimage_hash: string = row.preimage_hash;
+                            let preimage_distance: number = row.preimage_distance;
+                            let height_preimage = new Hash();
+
+                            // Hashing preImage
+                            if (row.preimage_hash &&
```
Comment (MichaelKim20) [inline src/Stoa.ts]: In exceptional cases, http status 404 should be returned.

File: src/Stoa.ts [hunk 4]
```diff
@@ -45,15 +45,30 @@ class Stoa {
...
+                                row.preimage_distance >= (height - row.enrolled_at))
+                            {
+                                height_preimage.fromHexString(preimage_hash);
+                                let cnt: number = (height - row.enrolled_at);
+                                for (let i = 0; i < cnt; i++)
+                                {
+                                    height_preimage.hash(height_preimage.buffer.slice());
+                                }
+                            }
```
Comment (MichaelKim20) [inline src/Stoa.ts]: https://github.com/bpfkorea/agora/blob/e7f8eeb55de9c914a6ba04b2c50b948b5e1ed0d4/source/agora/consensus/EnrollmentManager.d#L515-L532
I recommend you to check the condition and result value again.

**Review by MichaelKim20 — CHANGES_REQUESTED** [review]
I recommend adding the `/preimage_received` test at `Stoa.test.ts` to the first commit.
It is also recommended that you add a test for `/validators` in the second commit.
I recommend you to check if the returned preimage is the same as the expected value.

File: src/Stoa.ts (hunk 3 above)
Comment (MichaelKim20) [reply src/Stoa.ts]: I recommend finding a way to deal with exceptional circumstances.

**Comment by TrustHenry** [comment]
@MichaelKim20 Thanks for your review.
I'm going to apply @MichaelKim20  suggestion and switch it to a review.

File: src/Stoa.ts [hunk 5]
```diff
@@ -159,6 +161,42 @@ class Stoa {
...
+            if (req.body.pre_image == undefined)
+            {
+                res.status(400).send("Missing 'preImage' object in body");
+            }
+
+            try {
+                if (typeof req.body.pre_image === "string")
+                    pre_image = JSON.parse(req.body.pre_image);
+                else
+                    pre_image = req.body.pre_image;
```
Comment (Geod24) [inline src/Stoa.ts]: Why is this needed ?

File: tests/Storage.test.ts [hunk 6]
```diff
@@ -293,4 +294,111 @@ describe('LedgerStorage', () =>
...
+
+        putAllBlockData(ledger_storage, (err2: Error | null) =>
+        {
+            assert.ok(!err2, err2?.message);
+
+            let height: number = 0;
+            ledger_storage.getEnrollments(height,
+                () =>
+                {
+                    it ('Update preImage normal Test', (doneIt: () => void) =>
```
Comment (Geod24) [inline tests/Storage.test.ts]: Note on naming: The function named `it` should be used as the first word of the sentence (but not put in the string). See [the example](https://mochajs.org/#getting-started).
So in this case, `"It " ~ "Update preImage normal Test"` is not quite a sentence.
Something like `it('tests that sending a pre-image with a distance of 6 works')` would be more appropriate.

**Review by Geod24 — CHANGES_REQUESTED** [review]
There is no test for pre-images received after re-enrollment.

File: src/Stoa.ts (hunk 5 above)
Comment (TrustHenry) [reply src/Stoa.ts]: It was maintained for compatibility with the previous test, Agora History Push.
But now It doesn't need it
I will change it.
`pre_image = req.body.pre_image;`

**Comment by Geod24** [comment]
This needs to be rebased.

File: tests/Storage.test.ts (hunk 6 above)
Comment (TrustHenry) [reply tests/Storage.test.ts]: I improved the wording.

File: tests/SampleData.test.ts [hunk 7]
```diff
@@ -287,3 +287,10 @@ export const sample_data =
             ]
         }
     ];
+
+export const sample_preImageInfo =
+    {
+        "enroll_key": "0x210b66053c73e7bd7b27673706f0272617d09b8cda76605e91ab66ad1cc3bfc1f3f5fede91fd74bb2d2073de587c6ee495cfb0d981f03a83651b48ce0e576a1a",
+        "hash": "0x4869b90d82af612dac15b6f152700b2e0f0b4a198fa09d83853d4ac3be4032b051c48806692b37776534f2ae7b404c9221ae1c9616fe50e3585d63e607d0afc6",
+        "distance": 6
+    }
```
Comment (MichaelKim20) [inline tests/SampleData.test.ts]:
```suggestion
    };
```

File: src/modules/storage/LedgerStorage.ts [hunk 8]
```diff
@@ -285,6 +285,78 @@ export class LedgerStorage extends Storages
         });
     }
 
+    /**
+     * Update a preImage to database
+     * @param data a PreImageInfo
+     * @param onSuccess This function will be called when
+     * the database was finished successfully
+     * @param onError This function will be called when
+     * an error occurred.
```
Comment (MichaelKim20) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
```

**Review by MichaelKim20 — CHANGES_REQUESTED** [review]
I found two things that need to be corrected.

File: src/modules/storage/LedgerStorage.ts [hunk 9]
```diff
@@ -248,7 +248,7 @@ export class LedgerStorage extends Storages
                     [
                         block.header.height.value,
                         0,
-                        '0x0000000000000000',
+                        Hash.NULL,
```
Comment (MichaelKim20) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                        enroll.random_seed,
```
Isn't this right now?

Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: Yes, It was previously changed to random_seed, but Hash.NULL was left for a while.

**Comment by TrustHenry** [comment]
@MichaelKim20 Thank you for your review. I applied your suggestions.
@Geod24 Tests for pre-images received after re-enrollment has been added.
Ready for review.

**Comment by TrustHenry** [comment]
UpdatePreimage promise syntax has been simplified.
Simplify 'Hashing preImage' comparison condition

File: tests/Stoa.test.ts [hunk 10]
```diff
@@ -17,7 +17,13 @@ import express from "express";
 import axios from "axios";
 import * as http from "http";
 import URI from "urijs";
-import {sample_data} from "./SampleData.test";
+import {sample_data,
+        sample_preImageInfo,
+        sample_reEnroll_preImageInfo
+       } from "./SampleData.test";
+import { crypto_stream_NONCEBYTES } from 'sodium-native';
```
Comment (MichaelKim20) [inline tests/Stoa.test.ts]:
```suggestion
```
I just found out that this is not used.

**Review by MichaelKim20 — APPROVED** [review]
LGTM

File: tests/Stoa.test.ts (hunk 10 above)
Comment (TrustHenry) [reply tests/Stoa.test.ts]: Nice catch!

**Comment by TrustHenry** [comment]
@Geod24 ping

File: tests/Storage.test.ts [hunk 11]
```diff
@@ -345,4 +346,111 @@ describe('LedgerStorage', () =>
...
+                                    doneIt();
+                                });
+                        })
+                        .catch((err2) => {
+                            assert.ok(!err2, err2);
+                            doneIt();
+                        });
+                    });
+
+                    it ('Preimage distance Greater Than ValidatorCycle Test', (doneIt: () => void) =>
```
Comment (Geod24) [inline tests/Storage.test.ts]: Name is less than ideal, remember my advice during the previous round of review

**Comment by Geod24** [comment]
@TrustHenry : The review I provided was not addressed.

> There is no test for pre-images received after re-enrollment.

File: tests/Stoa.test.ts [hunk 12]
```diff
@@ -118,12 +124,144 @@ describe ('Test of Stoa API Server', () =>
...
+        enrollment.utxo_key =
+            "0x210b66053c73e7bd7b27673706f0272617d09b8cda76605e91ab66ad1cc3bfc1f3f5fede91fd74bb2d2073de587c6ee495cfb0d981f03a83651b48ce0e576a1a";
+        enrollment.enroll_sig =
+            "0x0c48e78972e1b138a37e37ae27a01d5ebdea193088ddef2d9883446efe63086925e8803400d7b93d22b1eef5c475098ce08a5b47e8125cf6b04274cc4db34bfd";
+        enrollment.random_seed =
+            "0xe0c04a5bd47ffc5b065b7d397e251016310c43dc77220bf803b73f1183da00b0e67602b1f95cb18a0059aa1cdf2f9adafe979998364b38cd5c15d92b9b8fd815";
+        block.header.enrollments.push(enrollment);
+
+        // put the re-enrollment
+        await stoa.ledger_storage.putEnrollments(block);
```
Comment (TrustHenry) [inline tests/Stoa.test.ts:203]: puted the re-enroll here And

File: tests/Stoa.test.ts [hunk 13]
```diff
@@ -118,12 +124,144 @@ describe ('Test of Stoa API Server', () =>
...
+        response = await client.get (uri9.toString());
+        console.log(response.data);
+        assert.strictEqual(response.data.length, 0);
+
+        // push the re-enroll's preImage
+        let uri10 = URI(host)
+        .port(port)
+        .directory("preimage_received");
+        response = await client.post (uri10.toString(), {pre_image: sample_reEnroll_preImageInfo});
+        assert.strictEqual(response.status, 200);
```
Comment (TrustHenry) [inline tests/Stoa.test.ts:251]: Test for pre-images received after re-enrollment.

File: tests/Storage.test.ts (hunk 11 above)
Comment (TrustHenry) [reply tests/Storage.test.ts]: Thanks for advice.
I change the wording.
`it('tests that sending a pre-image with a distance of 1008 works')`

**Comment by TrustHenry** [comment]
> @TrustHenry : The review I provided was not addressed.
> 
> > There is no test for pre-images received after re-enrollment.

@Geod24 
I did the test 'pre-images received after re-enrollment'
Is there a problem or insufficient test method?

File: tests/Stoa.test.ts [hunk 14]
```diff
@@ -118,12 +124,140 @@ describe ('Test of Stoa API Server', () =>
...
+        let block = new Block();
+        let height = new Height();
+        let enrollment = new Enrollment();
+        height.value = 1008;
+        block.header.height = height;
+
+        // re-enrollment
+        enrollment.cycle_length = 1008;
+        enrollment.utxo_key =
+            "0x210b66053c73e7bd7b27673706f0272617d09b8cda76605e91ab66ad1cc3bfc1f3f5fede91fd74bb2d2073de587c6ee495cfb0d981f03a83651b48ce0e576a1a";
```
Comment (Geod24) [inline tests/Stoa.test.ts:195]: Where did this validator previously enroll ? I can't find this hash anywhere.

Comment (TrustHenry) [reply tests/Stoa.test.ts:195]: Here it is.
https://github.com/bpfkorea/stoa/blame/v0.x.x/tests/SampleData.test.ts#L29

Comment (TrustHenry) [reply tests/Stoa.test.ts:195]: It's registered here.
https://github.com/bpfkorea/stoa/blob/51c52171fc60062235cb74588a6dd6c824c5c77e/tests/Stoa.test.ts#L79

**Comment by TrustHenry** [comment]
@Geod24 ping

**Comment by TrustHenry** [comment]
Rebase done.

**Review by Geod24 — APPROVED** [review]
